### DISCUSSION TITLE
ConnectionPool questions

### DISCUSSION THREAD

**Discussion by 91DarioDev** [psycopg/psycopg#321] [Q&A] [answered]
Hello, I read the page about ConnectionPool. ( https://www.psycopg.org/psycopg3/docs/advanced/pool.html ) Unfortunately it lacks of async snippets. Should the ConnectionPool be awaited?

Also why in those snippets after acquiring the connection you immediately use execute() method instead of using before cursor()?

It would be perfect to have a snippet usage of an async ConnectionPool 

Thanks

**Comment by dvarrazzo** [comment, marked as answer]
Constructors cannot be async, so we can't await on the class creation. The methods `open()` and `close()` are async instead. Connections are actually started by default at class creation, but you can make them start on `await open()` instead, by passing the parameter `open=False` to the constructor.

The [`AsyncConnectionPool` API docs](https://www.psycopg.org/psycopg3/docs/api/pool.html#the-asyncconnectionpool-class) show which methods are async.

Psycopg 3, unlike 2, has a [`Connection.execute()` method](https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.execute), which is [equivalent to creating a default cursor and executing on it](https://www.psycopg.org/psycopg3/docs/basic/usage.html#shortcuts). You can still create cursors explicitly, if it is better for your use cases.

Thank you for your suggestion to add examples of async usage for the pool: we will review the docs ;)

**Reply by 91DarioDev** [reply]
Ah thank you. So if I have understood well in case I use asyncio I don't have to use psycopg_pool.ConnectionPool but psycopg_pool.AsyncConnectionPool. I should call it without await and assign it to a global variable at the start of the script. Methods of AsyncConnectionPool should be awaited instead.
Is it right? Thank you very much

**Reply by dvarrazzo** [reply]
Yes, that's the right pattern.

If your framework provides startup and shutdown events (such as [FastAPI's](https://fastapi.tiangolo.com/advanced/events/#events-startup-shutdown)) you can create a global instance of the pool with `open=False` and call the `open()` and `close()` methods in these events handlers. See [pool life cycle](https://www.psycopg.org/psycopg3/docs/advanced/pool.html#pool-life-cycle) for an example.

**Reply by 91DarioDev** [reply]
Than you very much all clear! :)
